### PR TITLE
fix: fail fast on a nonexistent or unavailable ha_call_service/ha_bulk_control target

### PR DESCRIPTION
**Pull request by swissmo**
## Summary

- `ha_call_service` and `ha_bulk_control`'s single-entity write paths used to burn the full 10s state-change confirmation wait for a target that **does not exist**, even though such a target can never emit a confirming `state_changed` event. Both the component (`_call_service_prep`/`_bulk_call_service_prep`) and the legacy REST fallback (`tools_service.py`) now fail fast on a settled nonexistent target instead — no dispatch, no wait. A shared `is_single_entity_target()` predicate keeps this scoped to genuine single-entity lookups: a comma-separated multi-target or a Home Assistant magic broadcast target (`entity_id="all"`/`"none"`) is never fast-failed, since a states-endpoint miss for either proves nothing about whether the underlying call can succeed.
- The component already captures each target's pre-dispatch state for free (a synchronous in-memory `hass.states.get()` read); a `_confirmable_entity_ids()` helper excludes a target whose captured state is `None` (nonexistent) from the confirmation wait entirely, so the result comes back near-instantly with a structured `ENTITY_NOT_FOUND` instead.
- An **`"unavailable"`** target is handled differently, deliberately, on *both* paths: unlike a nonexistent entity, it can legitimately reconnect *during* the dispatch itself (e.g. a device that wakes on the very command being sent) — so the call always dispatches, and the target stays eligible to confirm. It's only reported as the distinct `ENTITY_UNAVAILABLE` once confirmation genuinely lapses AND a fresh post-dispatch state check shows it's still unavailable (not removed from the state machine entirely, and not reconnected to some state other than the exact expected hint) — gated on `not confirmed` in the component/bulk paths, and on a live re-check after a genuine `wait_for_state_change` timeout in the legacy path. Because this raise only ever fires *after* the command already reached Home Assistant, it carries `context["dispatched"] = True` and says so explicitly, so a caller doesn't misread it as "never sent" and retry a non-idempotent service into a double-apply.
- `ha_bulk_control`'s component path already detected `ENTITY_NOT_FOUND` post-hoc (after the full wait) from the same captured pre-state; this PR makes that detection skip the wait too, and adds the same `ENTITY_UNAVAILABLE`-after-genuine-lapse distinction (including the post-dispatch-state and removed-mid-dispatch checks) it was missing.

Every case raises the existing structured `ENTITY_NOT_FOUND`/`ENTITY_UNAVAILABLE` `ToolError` — a normal MCP tool result the caller can read and recover from (search for the right entity, retry, or ask the user), never a transport-level failure. Successful calls and genuine confirmation lapses on real, available targets are unchanged: same bounded wait, same `partial`/timeout shape.

### Root cause of the reported hang

A local-LLM voice pipeline (small model, short client-side deadline) invented `light.cuisine`, dispatched `turn_on`, and the client's own deadline expired mid-wait before the (eventually correct, but 10s-late) tool result could be delivered — so the model never learned the target was wrong and couldn't search/retry. The already-shipped `SessionDisconnectLogFilter` correctly demotes the resulting late-response log noise to a WARNING; this fixes the actual pointless wait underneath it.

### Compatibility

Bumps the component to 2.1.1 — behavior-only, no wire-schema change, so no `MIN_COMPONENT_VERSION` gate is needed. The component's wire envelope keeps the exact same shape for this case (a missing/unavailable single target still produces `confirmed=false, partial=true`, never `partial=false`) — only the *Python consumer's* handling of that envelope changed (raising a structured error instead of a generic warning for the two settled cases). A released (pre-this-PR) server reading component 2.1.1's envelope still sees its familiar `partial` shape and warns as it always did; an older component talking to the new server still gets the correct, if slower (full-timeout), classification.

## Test plan

- [x] New unit tests across every affected surface (legacy REST, component single-call, component bulk; both routing-level synthetic fixtures and cross-seam real-component-prep tests) for: fast-fail with zero confirmation listener registered for a nonexistent target; a comma-separated or magic-broadcast (`"all"`) target is never fast-failed as not-found; an "unavailable" target that reconnects to the expected state is reported a normal success; one that reconnects to some *other* state is neither a false success nor a false `ENTITY_UNAVAILABLE`; one removed mid-dispatch reports `ENTITY_NOT_FOUND`, not `ENTITY_UNAVAILABLE`; a genuine confirmation lapse (still unavailable, or a real/available target) reports the right shape with `dispatched: True` context where applicable; a mixed idempotently-confirmed + missing target does not wait the full timeout; a transient fetch error still dispatches (best-effort, unchanged); an empty `transitions` list (post-dispatch formatting failure after a successful dispatch) is not misread as proof of absence.
- [x] Dedicated unit test classes for the two new shared helpers (`is_single_entity_target`, `create_entity_unavailable_after_dispatch_error`).
- [x] Manifest/component version parity test updated for the 2.1.1 bump.
- [x] Full unit suite run locally (`pytest src/unit/ -n auto`, 11.5k+ tests) — no regressions versus unmodified `master` (a set of pre-existing, unrelated Windows-environment failures reproduce identically on both).
- [x] `ruff check` / `ruff format` clean on all changed files; `mypy` clean on the changed server-side modules.
- [x] CI-caught E2E regression (comma-separated multi-target 404 misread as not-found) fixed and pinned with a unit regression test.
- [x] All CI checks green; every automated review thread and outside-diff finding, and a full human maintainer review (4 major findings, live-verified against a real HA instance), addressed and replied to (see PR comments for the full trail).

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved confirmations for temporarily unavailable entities that reconnect during operations.
  * Added clearer errors for nonexistent or persistently unavailable entities, including dispatch status.
  * Prevented inconclusive or incomplete results from being misclassified as missing entities.
  * Improved mixed-target, broadcast, and bulk-operation reporting, including partial results and confirmation status.

* **Chores**
  * Updated the component version to 2.1.1.

### CODE REVIEW

**Comment by ghhamcp** [comment]
@codex review — apply the review criteria in .gemini/styleguide.md in addition to AGENTS.md guidance

File: custom_components/ha_mcp_tools/websocket_api.py [hunk 1]
```diff
@@ -5695,22 +5700,30 @@ async def _call_service_prep(
...
     pre = {eid: _state_as_dict(_state_get(hass, eid)) for eid in entity_ids}
+    # Only a target whose pre-dispatch state can possibly report a confirming
+    # event is worth waiting on (see ``_confirmable_entity_ids``) — a target
+    # absent from the state machine or already "unavailable" can never emit one,
+    # so waiting on it is certain to burn the full ``timeout`` for no new
+    # information. Excluding it turns that into an immediate, precise result
+    # (the server reads the ``None`` / "unavailable" ``old_state`` straight off
+    # the transition) instead of a pointless wait.
+    confirmable_entity_ids = _confirmable_entity_ids(entity_ids, pre)
+    should_confirm = bool(wait and confirmable_entity_ids)
```
[the remark this answers is not in the document]
Comment (swissmo) [reply custom_components/ha_mcp_tools/websocket_api.py]: ✅ Fixed in f16be15b. `_await_confirmation` now receives `confirmable_entity_ids` instead of the full `entity_ids` set (and is only called at all when `evt is not None`, i.e. something was actually registered to wait on). Added a direct regression test against the real component prep (`test_mixed_idempotent_and_missing_target_does_not_wait_full_timeout` in `test_component_call_service_contract.py`) asserting on wall-clock time, not just the outcome.

File: custom_components/ha_mcp_tools/websocket_api.py [hunk 2]
```diff
@@ -6155,19 +6192,25 @@ def _bulk_op_record(
...
     return {
         "domain": op["domain"],
         "service": op["service"],
         "service_data": op.get("service_data") or {},
         "entity_ids": entity_ids,
+        "confirmable_entity_ids": confirmable_entity_ids,
         "expected_by_entity": dict.fromkeys(entity_ids, expected_state),
-        "should_confirm": bool(wait and entity_ids),
-        "pre": {eid: _state_as_dict(_state_get(hass, eid)) for eid in entity_ids},
+        "should_confirm": bool(wait and confirmable_entity_ids),
```
[the remark this answers is not in the document]
Comment (swissmo) [reply custom_components/ha_mcp_tools/websocket_api.py]: ✅ Fixed in f16be15b. `should_confirm` is restored to intent-level (`bool(wait and entity_ids)`, unchanged from before this PR) instead of being narrowed to the confirmable subset — `confirmable_entity_ids` now only scopes which targets the component actually registers a listener / waits on, never whether confirmation was requested at all. So a null-prestate op with `validate_first=False` still lands on `confirmed=False, partial=True` exactly as before (just without burning the wait), and `_map_component_op_result`'s existing not-found/unavailable mapping (still gated on `validate_first`) is untouched. `test_validate_first_false_allows_null_prestate_op` (pre-existing) still passes unmodified with this fix.

File: custom_components/ha_mcp_tools/websocket_api.py [hunk 3]
```diff
@@ -5764,6 +5777,24 @@ async def _call_service_prep(
...
+    ``state_changed`` for this dispatch: HA no-ops a service call for an entity
+    id that matches nothing, and an unavailable entity has no live integration
+    to act on the command. Excluding either from the confirmation set lets the
+    server read the certain outcome straight off the transition's ``old_state``
+    instead of burning the full wait to learn nothing new.
+    """
+    return [
+        eid
+        for eid in entity_ids
+        if pre.get(eid) is not None and pre[eid].get("state") != "unavailable"
```
[the remark this answers is not in the document]
Comment (swissmo) [reply custom_components/ha_mcp_tools/websocket_api.py]: ✅ Fixed in f16be15b — good catch. `_confirmable_entity_ids` no longer excludes `"unavailable"` targets, only `None` (nonexistent) ones: an unavailable entity stays IN the confirmation wait, so a genuine reconnect-and-transition during the blocking dispatch is still captured and reported as a normal success. `ENTITY_UNAVAILABLE` is now only raised after a genuine confirmation lapse (gated on `not confirmed` in both `tools_service.py`'s `_build_component_call_response` and `device_control.py`'s `_map_component_op_result`), never upfront. Added `test_unavailable_entity_still_confirms_on_reconnect` (single-call) and `test_unavailable_entity_in_batch_confirms_on_reconnect` (bulk) to pin exactly this.

File: custom_components/ha_mcp_tools/websocket_api.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (swissmo) [reply custom_components/ha_mcp_tools/websocket_api.py]: ✅ Fixed in f16be15b. This was a real consequence of the same `should_confirm` narrowing CodeRabbit and your other comment flagged — restoring `should_confirm` to intent-level means a missing/unavailable single target now produces `confirmed=false, partial=true` (never `partial=false`), which is exactly the shape the released server's `_build_component_call_response` already handles (it warns on `partial`). So an old server talking to component 2.1.1 still gets its familiar partial-success warning for this case, not a silent unwarned success — no version-skew gap remains, and no `MIN_COMPONENT_VERSION` bump is needed.

File: custom_components/ha_mcp_tools/websocket_api.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (swissmo) [reply custom_components/ha_mcp_tools/websocket_api.py]: 📝 Not changing this one — I don't think it's a live risk here, though flagging the reasoning in case I'm missing something.

`_state_as_dict`'s `except Exception: return None` branch is marked `# pragma: no cover - defensive" in the existing code (predates this PR) and `_call_service_prep`'s own docstring already documents why: "HA core enforces JSON-serializable state attributes for its own REST/WS/recorder APIs, so a state that reached the component already serializes." A real, existing entity whose `.as_dict()" raises would require attribute data that violates that core invariant — the codebase already treats this as an unreachable defensive branch, not a realistic case to design a fast-fail policy around.

Threading a separate existence signal (independent of `_state_as_dict`'s serialization outcome) through to the consumer would need a wire-schema change to distinguish it from genuine absence, which feels disproportionate for a case core's own guarantees say can't happen in practice. Happy to revisit if there's a concrete way this fires that I'm not seeing.

**Review by swissmo — COMMENTED** [review]
Addressed all review feedback in f16be15b (and fixed a real CI-caught regression alongside it):

**CodeRabbit** — `_await_confirmation` now receives `confirmable_entity_ids`, not the full `entity_ids`; an idempotently-confirmed valid target mixed with an excluded one no longer waits out the full timeout. New direct regression test asserts on wall-clock time.

**Codex (P1, old-server compatibility)** — `should_confirm` is restored to intent-level (`bool(wait and entity_ids)`); `confirmable_entity_ids` now only scopes what the component actually waits *on*, never whether confirmation was requested. A missing/unavailable single target now produces `confirmed=false, partial=true` (never `partial=false`), so a `validate_first=False` caller still gets its dispatched-but-unconfirmed signal, and an old server talking to component 2.1.1 still sees its familiar partial-warning shape.

**Codex (P2, missed reconnect)** — `"unavailable"` is no longer excluded from the confirmation wait like a nonexistent target: it can legitimately reconnect and transition during the blocking dispatch, so it stays in the wait and is only reported `ENTITY_UNAVAILABLE` after a genuine confirmation lapse (gated on `not confirmed`). New tests pin both the reconnect-confirms and genuine-lapse cases, for both the single-call and bulk paths.

**Codex (P2, serialization ambiguity)** — declined with reasoning (reply on the thread): the `_state_as_dict` failure branch is pre-existing, marked `pragma: no cover - defensive`, and core's own JSON-serializability guarantees make it practically unreachable for a real entity.

**Real regression caught by CI's E2E suite** — a comma-separated `entity_id` ("light.a,light.b", a valid multi-target for the service-call payload) 404s against `/api/states/<id>` even when both individual entities are real, since that endpoint has no comma syntax. The legacy fast-fail now skips entirely for a multi-target `entity_id`, matching the component route's own comma exclusion. Added a unit regression test reproducing the exact failure mode; this was the sole root cause behind every failing E2E/HAOS job.

Also fixed the `ruff format` gap (two test files weren't reformatted before the first commit) and mirrored the `confirmable_entity_ids`/`should_confirm` correction into the bulk component path, which had a newly-reachable crash (`op["evt"].wait()` on a `None` evt) that I caught while making the fix.

All 5 threads replied to and resolved. Full unit suite (11.5k+ tests) green locally; only pre-existing, unrelated Windows-environment failures remain (verified identical on unmodified `master`).

**Comment by swissmo** [comment]
Re: CodeRabbit's outside-diff finding on \`_raise_for_missing_or_unavailable_target\` (empty \`transitions\` vs a present row with null \`old_state\`) — fixed in 91e44c70, exactly as proposed. Also found and fixed the identical pre-existing pattern in \`device_control.py\`'s \`_map_component_op_result\` (the bulk path carries the same ambiguity via \`_dispatched_unconfirmed_bulk_result\`, predates this PR, wasn't previously exercised by a test). Added a regression test for each (single-call and bulk) confirming a dispatched-but-unconfirmed result with empty \`transitions\` falls through to the generic partial/timeout wording instead of a false \`ENTITY_NOT_FOUND\`.

**Comment by ghhamcp** [comment]
@codex review — apply the review criteria in .gemini/styleguide.md in addition to AGENTS.md guidance

File: src/ha_mcp/tools/tools_service.py [hunk 4]
```diff
@@ -1465,6 +1500,51 @@ def _component_verified_state(
...
+                if isinstance(t, dict) and t.get("entity_id") == entity_id
+            ),
+            None,
+        )
+        if transition is None:
+            return
+        old_state = transition.get("old_state")
+        if old_state is None:
+            raise_tool_error(create_entity_not_found_error(entity_id))
+        if isinstance(old_state, dict) and old_state.get("state") == "unavailable":
```
[the remark this answers is not in the document]
Comment (swissmo) [reply src/ha_mcp/tools/tools_service.py]: ✅ Fixed in 90637c7. Good catch — `_raise_for_missing_or_unavailable_target` (and the bulk mapper's mirrored check) now also requires `new_state` to still show `"unavailable"` (or be absent) before raising, not just `old_state`. A reconnect to some OTHER state than the exact expected hint — which the hint-filtered listener never captures, so the wait genuinely lapses — now correctly falls through to the generic partial/timeout wording instead of a false `ENTITY_UNAVAILABLE`. Added regression tests for both the single-call and bulk paths, simulating the reconnect via a direct live-state-store mutation (mirroring what `_post_state`'s fallback re-read would see) since the hint filter means no matching event ever fires.

**Comment by swissmo** [comment]
Re: CodeRabbit's outside-diff finding on the legacy path failing unavailable targets before dispatch — fixed in 7cb6d27, exactly as proposed. `_validate_entity_before_wait` keeps the single-target 404 fast-fail but no longer fast-fails on `"unavailable"`; the service call now dispatches normally (it might be exactly what reconnects the device). `_verify_state_change` classifies a genuine `wait_for_state_change` timeout with a live re-check: still unavailable → `ENTITY_UNAVAILABLE`; reconnected to some other state than the hint → falls through to the existing generic partial/timeout wording — mirroring the post-dispatch-state check already applied to the component path for the same reason. Rewrote the two affected legacy-path tests and added a reconnect-to-other-state regression test.

File: src/ha_mcp/tools/tools_service.py [hunk 5]
```diff
@@ -967,15 +1012,41 @@ async def _verify_state_change(
...
+        Any fetch failure (including a 404 -- the entity vanished) is NOT
+        proof of recovery, so it counts as still unavailable; only a fetched
+        state that is NOT ``"unavailable"`` counts as recovered.
+        """
+        try:
+            current = await self._client.get_entity_state(entity_id)
+        except Exception:
+            return True
+        current_state = current.get("state") if current else None
+        return current_state is None or current_state == "unavailable"
```
[the remark this answers is not in the document]
Comment (swissmo) [reply src/ha_mcp/tools/tools_service.py]: ✅ Fixed in dcbabaa, exactly as proposed. `_still_unavailable` now raises `ENTITY_NOT_FOUND` on a 404 (the entity was removed between the initial fetch and the re-check — a more specific outcome) and returns `False` (not proof of anything) for any other fetch failure or a response with no usable `state`, so those fall through to the generic verification-failed warning instead of a confidently-wrong `ENTITY_UNAVAILABLE`. Added regression tests for both branches.

File: tests/src/unit/test_ha_call_service_component_routing.py [hunk 6]
```diff
@@ -471,6 +635,103 @@ async def test_no_capability_uses_legacy_post() -> None:
...
+
+    with patch_ws(ws, tools_service), pytest.raises(ToolError) as exc:
+        await call_service(domain="light", service="turn_on", entity_id="light.a")
+
+    assert "ENTITY_UNAVAILABLE" in str(exc.value)
+    assert "light.a" in str(exc.value)
+    # Dispatch DID happen this time — the target may have reconnected because
+    # of it, so failing before the POST would deny it that chance.
+    assert len(client.call_service_calls) == 1
+    tools_service.wait_for_state_change.assert_called_once()
```
[the remark this answers is not in the document]
Comment (swissmo) [reply tests/src/unit/test_ha_call_service_component_routing.py:771]: ✅ Added in dcbabaa — `assert client.get_state_calls == 2` on both existing tests, plus two new regression tests for the 404-during-recheck and inconclusive-failure-during-recheck branches CodeRabbit's sibling comment flagged.

**Review by kingpanther13 — CHANGES_REQUESTED** [review]
Reviewed the full diff against master. CI is green and the three-way nonexistent / unavailable-reconnected / unavailable-lapsed split is carefully built — the structural fast-fail assertions (`bus.listeners == []`) are the right way to pin it, and the "reconnected to some OTHER state" case that most implementations get wrong is explicitly covered.

Four things need changing before this lands. Findings 1 and 2 are confirmed against a live Home Assistant 2026.8.3 instance running this PR's server build (component still 2.1.0, so the component path was exercised as an older component would see it).

## 1. `entity_id="all"` (HA's ENTITY_MATCH_ALL) is misclassified as a nonexistent entity on both paths

`ENTITY_MATCH_ALL = "all"` is a supported service target (`homeassistant/const.py`). It is not in the state machine, and `GET /api/states/all` 404s.

- Legacy: `_validate_entity_before_wait` (tools_service.py:968) skips the fast-fail only for a comma list, so `"," not in "all"` → the 404 raises ENTITY_NOT_FOUND *before* `self._client.call_service(...)`. The broadcast never dispatches. Pre-PR, `_capture_initial_state`'s broad `except` swallowed that 404 and the call went through.
- Component: `_confirmable_entity_ids` returns `[]` for `["all"]`, the dispatch still fires (HA does expand and execute it), then `_raise_for_missing_or_unavailable_target` sees `old_state is None` and raises ENTITY_NOT_FOUND — a landed write reported as a missing entity.

Verified live, using a domain with zero entities so the broadcast was a guaranteed no-op:

- `ha_call_service(domain=<empty domain>, service="turn_off", entity_id="all")` → `ENTITY_NOT_FOUND: Entity 'all' not found`. That domain's `turn_off` **is** registered, so the component's pre-dispatch `ServiceNotFound` guard passed — the ENTITY_MATCH_ALL dispatch went to HA and the server then reported the target as nonexistent.
- The same call with `verbose=True` (routes to legacy) → identical `ENTITY_NOT_FOUND`, and there the fast-fail provably precedes the service POST, so nothing reached HA at all.

On a domain that actually has entities, the first path executes the broadcast and reports "entity not found"; the second silently does nothing. The two paths now disagree on the same input, and one of them no-ops a call that used to work. `wait=False` is the only escape.

This is not a hypothetical shape: the repo's own fixture at tests/src/unit/test_saved_tools_persistence.py:321 is a saved tool whose body is `ha_call_service(domain="light", service="turn_off", entity_id="all")`.

Fix is the same shape as the comma exclusion already there — exclude `"all"` (or, more defensively, any target with no `"."`) in `_validate_entity_before_wait`, and mirror it in `_raise_for_missing_or_unavailable_target` and device_control.py's bulk mapping. Both surfaces: fixing only the legacy one leaves the component path reporting failure on a completed write. Please pin it with a regression test alongside `test_comma_multi_target_survives_404_on_states_endpoint`.

## 2. The ENTITY_UNAVAILABLE raise discards a write that already landed, and invites a retry of it

Both raise sites fire strictly after dispatch — legacy at tools_service.py:1016, inside `_verify_state_change`, which runs at :2023 after `call_service` returned and after `response` (including `service_response`) was assembled at :2005-2019; component at :1549 via `_build_component_call_response`, and the component always dispatches regardless of `_confirmable_entity_ids`.

For a nonexistent target that is right — HA no-ops it. For an "unavailable" target the command reached HA. Verified live: `turn_on` against an unavailable switch entity returned

```
ENTITY_UNAVAILABLE: Entity '<id>' exists but is currently unavailable
suggestions: [..., "Check if the device is powered on and connected", ...]
```

(the default suggestions from errors.py:179-183). The command was dispatched; nothing in that payload says so. `toggle` is in `_STATE_CHANGING_SERVICES`, so an agent that reads "check if the device is powered on" and retries double-applies a non-idempotent write — the D9 at-most-once boundary the component code defends at length. With `return_response=True` the `service_response` the caller asked for is discarded on the way out too.

The sibling bulk implementation at device_control.py:1090-1098 already attaches `err["service_call"] = service_call`; the single-call path is the odd one out. Minimum fix: put `"dispatched": True` in the `context=` of both raises and say so in the message and suggestions. Also worth reconciling with the contract stated at tools_service.py:1692-1695 ("success stays True; verification is never a failure"), which this PR makes untrue for one case without amending it — and with the PR description's compatibility argument, which describes a `partial=True` shape rather than the raise that shipped.

## 3. A removal mid-dispatch is classified differently on each path

tools_service.py:1593 and device_control.py:1075 both fold `new_state is None` into `still_unavailable`. `_post_state` returns `None` when the entity is gone from the state machine at post-dispatch time — removed, not unavailable — so those paths emit "Entity 'x' exists but is currently unavailable" (an assertion of existence that is false) with the wrong recovery suggestions.

dcbabaa fixed exactly this on the legacy path: `_still_unavailable`'s docstring says a 404 there "means the entity was REMOVED ... so it raises ENTITY_NOT_FOUND directly rather than being folded into it", with a test (`test_legacy_unavailable_target_removed_reports_not_found_not_unavailable`). The component and bulk paths still fold it. Split the `None` case out in both and mirror the test.

## 4. `_bulk_op_record`'s docstring states the opposite of the code, twice

websocket_api.py:6207-6211:

> `confirmable_entity_ids` excludes any target whose captured pre-state is `None` (nonexistent) or `"unavailable"` — [...] `should_confirm` is true only when the batch is waiting AND this op has at least one such target.

`_confirmable_entity_ids` (:5793) excludes only `None`, and its own docstring says "Deliberately NOT excluded: a target whose captured state is `"unavailable"`". `should_confirm` at :6230 is `bool(wait and entity_ids)` — the full list — and the inline comment six lines above it says exactly that. So the one docstring for the batch record documents the PR's load-bearing invariant backwards, in the place a maintainer looks first; someone "fixing" the code to match it would reintroduce the bug this PR exists to remove.

Two more comments carry the same inversion, plus one in device_control.py:

- tools_service.py:1977 — "A settled bad target (404 / `"unavailable"`) raises here, BEFORE the dispatch": an unavailable target deliberately does dispatch. Drop the `/ "unavailable"`.
- tools_service.py:1637 — "nonexistent, or already `"unavailable"` [...] never reaches `confirmed`": an unavailable target can reach `confirmed`, which is why the guard below is gated on `not confirmed`.
- device_control.py:1023 — "maps to a distinct ENTITY_UNAVAILABLE failure the same way": it is doubly gated (`not confirmed` AND `still_unavailable`), not the same way as the NOT_FOUND branch.

## Smaller, same round

`_still_unavailable` (tools_service.py:1036) returns a bare `False` on every non-404 failure with no logging in either arm. `HomeAssistantAuthError` is a sibling of `HomeAssistantAPIError`, not a subclass (rest_client.py:101/117), so a 401 lands in the bare `except Exception` and an expired token becomes "state change could not be verified within timeout" with no server-side trace. Its two siblings both log — `_validate_entity_before_wait` in this same diff, and `wait_for_state_change` in util_helpers.py, which logs a warning and re-raises the connection/auth pair.

`test_legacy_unavailable_recheck_transient_error_falls_through` asserts only that no "unavailable" warning appears, which also passes if `warnings` vanishes entirely. Assert the generic warning positively, as the three sibling tests do.

**Comment by swissmo** [comment]
@kingpanther13 Thank you for the thorough, live-verified review — this caught real bugs unit tests alone hadn't surfaced. Addressed all four findings plus the smaller ones in 6b0c541:

## 1. `entity_id="all"` misclassified as nonexistent — fixed on all three surfaces

New shared `is_single_entity_target()` predicate (`util_helpers.py`): false for a comma-joined multi-target OR a dotless string (covers `"all"`/`ENTITY_MATCH_ALL`, `"none"`/`ENTITY_MATCH_NONE`, and any future magic target the same way, rather than hardcoding by name). Gates:
- `_validate_entity_before_wait` (legacy) — the 404-on-"all" no longer fast-fails before dispatch; the broadcast reaches HA exactly like pre-PR.
- `_raise_for_missing_or_unavailable_target` (component, single-call) — a null `old_state` for "all" no longer raises `ENTITY_NOT_FOUND` on a call that actually landed.
- `_map_component_op_result` (bulk) — same guard. Worth noting for this one: `ha_bulk_control`'s explicit per-op format already rejects a dotless `entity_id` upstream (`_resolve_component_rows`'s own `"." not in entity_id` check, pre-existing, unrelated to this PR) as `ENTITY_INVALID_ID` via legacy fallback — so `"all"` specifically never reaches this function in bulk. The guard is still correct/defensive there and matters for the comma-multi-target case, which *does* reach it (added `test_batch_skips_comma_multi_target` to cover that reachable case instead).

Added `test_saved_tools_persistence.py`-style regression coverage: `test_comma_multi_target_survives_404_on_states_endpoint`'s sibling `test_legacy_skips_magic_broadcast_target` (legacy), `test_component_skips_magic_broadcast_target`/`_real_prep` (component, routing + cross-seam), `test_batch_skips_comma_multi_target` (bulk).

## 2. `ENTITY_UNAVAILABLE` now carries `dispatched: True`

New shared `create_entity_unavailable_after_dispatch_error()` (`errors.py`) — `context["dispatched"] = True`, and the message/details/suggestions now say explicitly that the command already went out and that retrying a non-idempotent service could double-apply it. Used by all three raise sites (legacy `_verify_state_change`, component `_raise_for_missing_or_unavailable_target`, bulk `_map_component_op_result`). Also reconciled the two stale "success stays True; verification is never a failure" / "never reaches confirmed" comments you flagged — see finding 4.

On the PR description: fixed to distinguish the component's wire-envelope shape (still `partial=true`, unchanged — that's what makes the old-server compatibility argument hold) from what the Python consumer now does with it (raises, for the specific proven-not-found/still-unavailable cases).

## 3. Removal mid-dispatch — split from "still unavailable" on component + bulk

`new_state is None` (post-dispatch live re-read finds nothing) now raises `ENTITY_NOT_FOUND` instead of being folded into `still_unavailable`, on both the component (`_raise_for_missing_or_unavailable_target`) and bulk (`_map_component_op_result`) paths — matching what `_still_unavailable` already did for its own 404-during-recheck case on the legacy path. Regression tests: `test_component_reports_not_found_when_removed_mid_dispatch` (routing + cross-seam) and `test_batch_reports_not_found_when_removed_mid_dispatch`.

(Also found my own test fixture bug while fixing this — `_unavailable_result`'s `new_state` was `None`, which after this fix meant it was accidentally exercising the *new* removed-mid-dispatch branch instead of genuine still-unavailable. Fixed the fixture to a real `"unavailable"` state dict.)

## 4. Docstring/comment inaccuracies — all four fixed

- `_bulk_op_record` (the one you flagged as most concerning, documenting the invariant backwards): now correctly states `confirmable_entity_ids` excludes ONLY `None`, and `should_confirm` stays intent-level over the full `entity_ids`.
- `tools_service.py` legacy-fallback comment: dropped the stale `/ "unavailable"` — only a 404 fast-fails before dispatch now.
- `_build_component_call_response`'s "never reaches confirmed" comment: corrected — unavailable targets CAN reach confirmed (that's the point of keeping them in the wait).
- `device_control.py`'s `_map_component_op_result` docstring: now notes the unavailable branch is doubly-gated (not confirmed AND still-unavailable-post-dispatch), unlike the not-found branch.

## Smaller

- `_still_unavailable`'s non-404 branches now `logger.warning` (matching its siblings) instead of silently swallowing failures — an expired token (`HomeAssistantAuthError`, a sibling of `HomeAssistantAPIError`, not a subclass) now leaves a trace instead of reading as a generic timeout.
- `test_legacy_unavailable_recheck_transient_error_falls_through` now asserts the generic warning positively, not just the absence of "unavailable".

Full local unit suite green (11.5k+ tests, only the same pre-existing unrelated Windows-environment failures), `ruff`/`mypy` clean, new coverage for every finding across every affected surface.

**Review by kingpanther13 — COMMENTED** [review]
Re-reviewed the delta (6b0c541). All four findings and both smaller ones are correctly addressed — I verified each against the code rather than the summary.

1. Magic broadcast targets — `is_single_entity_target` (util_helpers.py:56) requires a dot and no comma, and is applied at all three surfaces: `_validate_entity_before_wait` (legacy fast-fail), `_raise_for_missing_or_unavailable_target` (component single-call), and `_map_component_op_result`'s validate block (bulk). ENTITY_MATCH_NONE ("none") is covered too — I confirmed both sentinels are in homeassistant/const.py. Deriving it from the `domain.object_id` shape rather than a name list is the right call.

   One thing worth stating explicitly, because it is a deliberate behavior choice and not obvious from the tests: `entity_id="all"` on the component path now returns `success: true, partial: true` plus the generic could-not-verify warning. That is correct — the broadcast did land and no single transition can confirm it — but it means an "all" call is reported exactly like a genuine confirmation lapse. Fine by me; flagging it so it is an intentional contract rather than an accident.

2. Post-dispatch ENTITY_UNAVAILABLE — `create_entity_unavailable_after_dispatch_error` carries `context["dispatched"] = True`, a `details` line saying the call already reached HA, and suggestions that name the non-idempotent-retry hazard. Used at all three raise sites. The docstring explaining why this is distinct from a bare unavailable report is the right place for that reasoning.

3. Removal mid-dispatch — both the component single-call path (tools_service.py:1632) and the bulk mapper (device_control.py:1099) now split `new_state is None` out to ENTITY_NOT_FOUND instead of folding it into "still unavailable", each with a regression test. The three paths agree now.

4. Docstrings — `_bulk_op_record` states the invariant correctly in both halves, and the three other inverted comments (tools_service.py:2010 call-site, :1663 pre-check, device_control.py:1023) are fixed. The `_raise_for_missing_or_unavailable_target` docstring's new three-bullet new_state table is a real improvement over the prose it replaced.

   `_still_unavailable` logs at WARNING in both inconclusive arms and calls out the HomeAssistantAuthError-is-a-sibling trap by name. The transient-recheck test now asserts the generic warning positively.

Outstanding, and the only thing between this and green:

The CodeQL Code Quality gate fails on 6b0c541 with 2 findings, both new and both from this commit — `py/implicit-string-concatenation-in-list` at src/ha_mcp/errors.py:428 and :430, the two wrapped literals inside the new helper's `suggestions=[...]`. CodeQL reads adjacent string literals in a list as a possible missing comma. It is a false positive, but this gate blocks. Either collapse each suggestion to a single line, or add the two entries to ALLOWLIST in scripts/codeql_quality_gate.py with a justification (the gate pins a suppression to rule + path + message + the flagged source text, so both lines need their own entry). `CodeQL Gate` failing is just the rollup of that job; everything else is green.

Happy to approve once that is fixed and CI is green — ping me and I will re-run the review against the new head.

**Comment by swissmo** [comment]
@kingpanther13 Thank you for re-reviewing and pinpointing this so precisely. Fixed in 7334ac1: switched both suggestion strings in \`create_entity_unavailable_after_dispatch_error\` to explicit \`+\` concatenation instead of implicit adjacent-literal concatenation — matches this same file's existing convention for the identical situation (\`_ADDON_AUTH_INVALID_SUGGESTIONS\`). No behavior change, no allowlist entry needed.

Also, thank you for confirming the \`entity_id="all"\` contract explicitly (partial=true + generic could-not-verify warning) — agreed that's the right, intentional behavior for a broadcast target where no single transition can confirm it, and good to have it stated as a deliberate contract rather than left implicit.

CI is running on 7334ac1 now — will confirm here once it's green.

**Review by kingpanther13 — APPROVED** [review]
CodeQL fix verified. Explicit `+` between the two wrapped literals in `create_entity_unavailable_after_dispatch_error`'s suggestions silences py/implicit-string-concatenation-in-list without an allowlist entry, and the resulting strings are byte-identical to before — no behavior change, and no suppression to keep current later. Merging master in rather than force-pushing was the right call; it kept the review history intact.

Only PR-owned file touched since my last pass is errors.py (2 lines). All checks pass, all review threads resolved.

Approving. Everything from both earlier rounds is addressed, and I re-verified each fix against the code:

- Broadcast/multi-target guard applied at all three surfaces via `is_single_entity_target`, covering both HA sentinels
- Post-dispatch ENTITY_UNAVAILABLE now carries `dispatched: True` plus retry-hazard wording at all three raise sites
- Removal-mid-dispatch split out to ENTITY_NOT_FOUND on the component and bulk paths, matching the legacy path, each with a regression test
- The `_bulk_op_record` docstring and the three sibling inverted comments state the invariant correctly
- `_still_unavailable` logs both inconclusive arms at WARNING and names the HomeAssistantAuthError sibling trap

Nice work on the live-verified `entity_id="all"` case in particular — that one had no test and no bot had caught it.
